torch/quantization/ns/graph_matcher.py

```diff
@@ -0,0 +1,315 @@
+            for arg in cur_node.args:
+                if isinstance(arg, Node):
+                    self.stack.append(arg)
+                # TODO(future PR): handle other arg types such as Tuple, etc
+
+            # skip observers, etc
+            if not self._is_matchable(cur_node):
+                continue
+
+            return cur_node
+
+        raise StopIteration
+
+    def _is_matchable(self, node: Node) -> bool:
+        if node.op == 'call_function':
+            return not (node.target in self.non_matchable_functions)
+        elif node.op == 'call_module':
+            assert isinstance(node.target, str)
+            # target_mod = getattr(self.gm, node.target)
+            target_mod = _getattr_from_fqn(self.gm, node.target)
+            return not \
+                any(isinstance(target_mod, t)  # type: ignore
+                    for t in self.non_matchable_modules)
+        else:
+            return False
+
+class GraphMatchingException(Exception):
+    """
+    Exception raised when two graphs cannot be matched.
+    """
+    pass
+
+def _node_a_related_to_b(
+    node_a: Node,
+    node_b: Node,
+    gm_a: GraphModule,
+    gm_b: GraphModule,
+    type_a_related_to_b: Set[Tuple[Callable, Callable]],
+) -> bool:
+    if node_a.op != node_b.op:
+        # for now, comparing call_module to call_function is not supported
+        # this can be added later if needed
+        return False
+
+    if node_a.op == 'call_function':
+        if node_a.target == node_b.target:
+            # nodes with equivalent targets always match (i.e. F.linear and F.linear)
+            return True
+        key = (node_a.target, node_b.target)
+        return key in type_a_related_to_b
+    elif node_a.op == 'call_module':
+        # for call_module, we need to look up the modules to do the type check
+        assert isinstance(node_a.target, str)
+        mod_a = _getattr_from_fqn(gm_a, node_a.target)
+        assert isinstance(node_b.target, str)
+        mod_b = _getattr_from_fqn(gm_b, node_b.target)
+        # modules with equivalent types always match (i.e. nn.Conv2d and nn.Conv2d)
+        if type(mod_a) == type(mod_b):
+            return True
+        key = (type(mod_a), type(mod_b))
+        return key in type_a_related_to_b
+    return False
+
+def _get_name_for_node_pair(
+    node_a: Node,
+    node_b: Node,
+) -> str:
+    if node_b.op == 'call_module':
+        assert isinstance(node_b.target, str)
+        return node_b.target
+    # for now, use node name.
+    # TODO(future PR): find a better solution
+    return node_b.name
+
+def _get_node_target_type(node: Node, gm: GraphModule) -> Optional[Callable]:
+    if node.op == 'call_function':
+        return node.target  # type: ignore
+    elif node.op == 'call_module':
+        assert isinstance(node.target, str)
+        mod = _getattr_from_fqn(gm, node.target)
+        return type(mod)
+    return None
+
+def get_matching_node_pairs(
```

Comment:
qq: have we explored whether we can use https://github.com/pytorch/pytorch/blob/master/torch/fx/subgraph_rewriter.py ?

Reply:
yes, let's ask the FX folks

Reply:
looks like it might take a few days to align on this, happy to update it in a future PR if there is a good fit.